.docs/contributing-book/src/documentation/project-quality/code-structure.md

```diff
@@ -0,0 +1,20 @@
+# Code Structure
+
+In order to structure your code there is only 1 rule that you must remember and that is that everything must always be ordered alphabetically - as long as there isn't a dependency error or a better name for a default field of an enum. The reason is that it provides a structure that almost everyone will be familiar with. This means that nobody has to learn how you have decided to structure your code and thus where to look for a variable, function, import etc. This makes it really easy to navigate your entire project and in turn navigate all projects without having to memorize the intricacies of each project.
+
+To be explicit
+
+- All dependencies must be declared in alphabetical order just below the declaration of the program type e.g. below `contract;`
+- All imports must be in alphabetical order and any modules within those imports must be in alphabetical order
+- All functions, structs, enums, fields inside structs and enums, parameters of a function, parameters in documentation etc. must be in alphabetical order in all files (this does not apply to variables inside a function)
```

Comment:
I agree that alphabetical ordering is best _within_ code structures (struct/enum fields, function params, etc.) but I think we should leave the order of those structures within the file to the discretion of the contributor, or at least soften the "must" requirement to a "should" (i.e. a good default to go with unless there's a compelling reason not to).

For example, it often makses sense to group functions by permission level - e.g. permissioned admin functions.

Reply:
Yes you can group things together, and the alphabetical thing annoys me at times too, nevertheless it does not counter the argument of increased cognitive load because you potentially have to familiarise yourself with the code each time you see it if you've not seen it in a while or if things have grown over time. If a new structure is proposed that is closer to grouping and is in similar cognitive load i.e. "learn this rule instead of alphabetical rule here" then I'll be fine with that otherwise it allows devs to throw things where ever any one of them sees fit and I rather not spend my attention trying to figure out how to traverse any new codebase or how things have changed.

Reply:
I'm not sure I follow that alphabetical order necessarily reduces cognitive load. It's a handy convention, but if I encounter code I'm not familiar with, it's rare for me to think "right, where can I find the functions that begin with 'J' " 🙃 

Happy to go with the consensus, once reached. But I do think this should be mulled over a bit, since it sounds like not all developers (from a sample size of 3 😁) approach this this same way.

I'm also keen to find a happy medium that respects the reviewer's time but also lets the developer work without feeling too constrained.

Reply:
My last comment was a nudge for a proposition of an alternative. Atm there are 2 devs that at least follow, but I don't know if they agree with, the convention. If an immediate proposition is not going to be made then there is no reason for this PR to sit here when it can be approved and updated in the future if a proposition comes in.

Reply:
Solidity has an established convention in terms of style & location of elements in a file.
It isn't all applicable here, as we have:
1.) `forc fmt` to enforce universal formatting
2.) good library support to avoid the need to dump everything needed by a contract into 1 file.

That said, establishing a convention about function order based on purity would be in line with what solidity does.
We could extend that to cover cases where the dev wants to use a single file structure, i.e: a small contract with 1 error, 1 event, 1 private(util) function and 1 data structure should not be forced to uses a 5-file project structure in my opinion.
For a more complex project, I really like the multi-module setup though, so I think allowing leeway for use of judgement makes sense.

Reply:
Can you provide more information regarding purity i.e. where do read only functions go, where do write only functions go, where do read/write functions go, what about the ordering of those functions within their own block of functions?
I think that structure is acceptable once a concrete structure is put into place.

Something that I'm not a fan of is grouping by concept e.g. "admin things here" because now I have to learn about each project to know where things are and there are a lot of stipulations and exceptions that would need to be listed as opposed to the 3 cases for purity in Sway.

I'm indifferent to content being placed in 1 file as long as it's not an actual "project" / "application" because that isn't an application imo. There's not enough code to warrant that as an app therefore I don't think that applies to the apps repo nor should that content be placed there.
There might be some outliers in early development e.g. the OTC swap predicate may not need additional functionality and can all fit in 1 file _at the moment_.

I'm fine with people structuring things in a consistent and logical way in a larger project. So far I have advocated for a `data_structures.sw` _file_ but that may be inappropriate for some projects and thus a directory of individual sway files might apply (as an example).

So, if information can be provided about the exact rules then I can come up with some text and update the PR.

Reply:
It sounds to me like there is a strong case for _not_ having stringent rules for ordering. I definitely acknowledge that there's a trade-off between cognitive overhead for the developer and the reviewer, but i'm hearing a preference amongst _most_ contributors for a little more freedom in this regard.

I would be in favor of dropping this requirement until there is some consensus. If there's no buy-in from contributors at large, it won't stick anyway.

Reply:
@Braqzen https://docs.soliditylang.org/en/v0.8.17/style-guide.html#order-of-functions

Reply:
The only relevant thing I see comes to the following ordering

1. `write`
2. `read`
3. `pure`

I think all `writes` require the `read` annotation as well so 

1. `write & read`
2. `read`
3. `pure`

for functions inside a contract. I don't really see the point of alphabetically ordering these. You could but I'm uncertain if that would be more useful than the amount of work needed to do this.

So, is the following acceptable:

1. variables / fields in data structures should be alphabetical
2. group data structures by module and alphabetically order the content there
3. when testing order the functions in the order of the requirements being tested
4. abi functions should be ordered as stated above (in terms of purity)
5. imports and dependencies should be alphabetical
6. document the interface in alphabetical order but the parameters themselves can be in any order in a function

If I've missed anything out then I might edit or add another message

Reply:
I think the above looks pretty good to me. 
For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.

Reply:
Awaiting confirmations from @simonr0204 @matt-user @bitzoic @SwayStar123 @supiket

Reply:
I think "getter" functions should be separated / put at the top/bottom of the abi impl block, it doesnt make sense to mix them into the functions that actually do stuff

Reply:
> I think the above looks pretty good to me. For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.

wait, the order of function _params_ in the abi and in the implementation need to match, right? The function signatures have to be identical.  i.e. `myfunc(a: u64, b: Address)` != `myfunc(b: Address, a: u64)` ?

Reply:
It does make sense. It depends on which perspective you impose on yourself. For some reason your choice is a separation between getters and non-getters rather than some other rule. This rule is "okay" but it allows a looser format than via purity.

Reply:
> > I think the above looks pretty good to me. For point # 6, I would probably prefer to document the interface params in the order they are declared in the function signature as opposed to alphabetical, but it's not a hill I need to die on. Happy to go with the majority here.
> 
> wait, the order of function _params_ in the abi and in the implementation need to match, right? The function signatures have to be identical. i.e. `myfunc(a: u64, b: Address)` != `myfunc(b: Address, a: u64)` ?

Yes, they must match however the documentation (the `///`) is what I was referring to. I want that to be alphabetical but I can live with the params being out of order although I would prefer them to be ordered. The `ABI` and implementation signatures must match though.

Reply:
ohh I see. Yeah that's fine.

#shipit